src/FrenetOptimalTrajectory/FrenetOptimalTrajectory.cpp

```diff
@@ -31,44 +35,230 @@ FrenetOptimalTrajectory::FrenetOptimalTrajectory(
     csp = new CubicSpline2D(x, y);
 
     // calculate the trajectories
-    calc_frenet_paths();
+    if (fot_hp->num_threads == 0) {
+        calc_frenet_paths();
+    } else { // if threading
+        vector<thread> threads;
+
+        // calculate how to split computation across threads
+        int num_di_iter = static_cast<int>(
+            (fot_hp->max_road_width_l + fot_hp->max_road_width_r) /
+            fot_hp->d_road_w);
+        num_di_iter = num_di_iter + 1; // account for the last index
+
+        int iter_di_index_range =
+            static_cast<int>(num_di_iter / fot_hp->num_threads);
+
+        for (int i = 0; i < fot_hp->num_threads; i++) {
+            if (i != fot_hp->num_threads - 1) {
+                threads.push_back(thread(
+                    &FrenetOptimalTrajectory::calc_frenet_paths_threaded, this,
+                    i * iter_di_index_range, (i + 1) * iter_di_index_range));
+            } else { // account for last thread edge case
+                threads.push_back(
+                    thread(&FrenetOptimalTrajectory::calc_frenet_paths_threaded,
+                           this, i * iter_di_index_range, num_di_iter));
+            }
+        }
+
+        // wait for all threads to finish computation
+        for (auto &t : threads) {
+            t.join();
+        }
+    }
 
     // select the best path
     double mincost = INFINITY;
-    for (FrenetPath* fp : frenet_paths) {
+    for (FrenetPath *fp : frenet_paths) {
         if (fp->cf <= mincost) {
             mincost = fp->cf;
             best_frenet_path = fp;
         }
     }
     auto end = chrono::high_resolution_clock::now();
-    double run_time = chrono::duration_cast<chrono::nanoseconds>(end - start).count();
+    double run_time =
+        chrono::duration_cast<chrono::nanoseconds>(end - start).count();
     run_time *= 1e-6;
-    //cout << "Planning runtime " << run_time << "\n";
+    // cout << "Planning runtime " << run_time << "\n";
 }
 
 FrenetOptimalTrajectory::~FrenetOptimalTrajectory() {
+    delete mu;
     delete csp;
-    for (FrenetPath* fp : frenet_paths) {
+    for (FrenetPath *fp : frenet_paths) {
         delete fp;
     }
 
-    for (Obstacle* ob : obstacles) {
+    for (Obstacle *ob : obstacles) {
         delete ob;
     }
 }
 
 // Return the best path
-FrenetPath* FrenetOptimalTrajectory::getBestPath() {
-    return best_frenet_path;
+FrenetPath *FrenetOptimalTrajectory::getBestPath() { return best_frenet_path; }
+
+/*
+ * Threaded version of calc_frenet_paths
+ * We parallelize on the outer loop, in terms of di
+ * Iterates over possible values of di, from start index to end index
+ * (exclusive). Then, computes the actual di value for path planning.
+ */
+void FrenetOptimalTrajectory::calc_frenet_paths_threaded(int start_di_index,
```

Comment:
I'm concerned that this method duplicates a lot of code. IMO it would be better if we had a single calc_frenet_path(int start_di_index, int end_di_index, multithreaded). If multithreaded is set to true then the method has to lock when adding a path, otherwise it doesn't.

Reply:
That makes sense.